specs/core/1_custody-game.md

```diff
@@ -536,252 +352,93 @@ def process_early_derived_secret_reveal(state: BeaconState, reveal: EarlyDerived
-                            if record.challenge_index == response.challenge_index), None)
-    if chunk_challenge is not None:
-        return process_chunk_challenge_response(state, response, chunk_challenge)
-
-    bit_challenge = next((record for record in state.custody_bit_challenge_records
-                          if record.challenge_index == response.challenge_index), None)
-    if bit_challenge is not None:
-        return process_bit_challenge_response(state, response, bit_challenge)
-
-    assert False
-```
-
-```python
-def process_chunk_challenge_response(state: BeaconState,
-                                     response: CustodyResponse,
-                                     challenge: CustodyChunkChallengeRecord) -> None:
-    # Verify chunk index
-    assert response.chunk_index == challenge.chunk_index
-    # Verify bit challenge data is null
-    assert response.chunk_bits_branch == [] and response.chunk_bits_leaf == Bytes32()
-    # Verify minimum delay
-    assert get_current_epoch(state) >= challenge.inclusion_epoch + MAX_SEED_LOOKAHEAD
-    # Verify the chunk matches the crosslink data root
-    assert is_valid_merkle_branch(
-        leaf=hash_tree_root(response.chunk),
-        branch=response.data_branch,
-        depth=challenge.depth,
-        index=response.chunk_index,
-        root=challenge.data_root,
-    )
-    # Clear the challenge
-    records = state.custody_chunk_challenge_records
-    records[records.index(challenge)] = CustodyChunkChallengeRecord()
-    # Reward the proposer
-    proposer_index = get_beacon_proposer_index(state)
-    increase_balance(state, proposer_index, Gwei(get_base_reward(state, proposer_index) // MINOR_REWARD_QUOTIENT))
+    assert bls_verify(malefactor.pubkey, hash_tree_root(epoch_to_sign), custody_slashing.malefactor_key, domain)
+
+    # Get the custody bit
+    custody_bits = attestation.custody_bits[custody_slashing.data_index]
+    committee = get_beacon_committee(state, attestation.data.slot, attestation.data.index)
+    claimed_custody_bit = custody_bits[committee.index(custody_slashing.malefactor_index)]
+    
+    # Compute the custody bit
+    computed_custody_bit = compute_custody_bit(custody_slashing.malefactor_key, custody_slashing.data)
+    
+    # Verify the claim
+    if claimed_custody_bit != computed_custody_bit:
+        # Slash the malefactor, reward the other committee members
+        slash_validator(state, custody_slashing.malefactor_index)
+        others_count = len(committee) - 1
+        whistleblower_reward = Gwei(malefactor.effective_balance // WHISTLEBLOWER_REWARD_QUOTIENT // others_count)
+        for attester_index in attesters:
+            if attester_index != custody_slashing.malefactor_index:
+                increase_balance(state, attester_index, whistleblower_reward)
+        # No special whisteblower reward: it is expected to be an attester. Others are free to slash too however. 
```

Comment:
This makes sense as there's no way to protect against front-running. Strange side-effect though, if several attesters produced incorrect custody bits, then some of them could collect the whistleblower rewards before being slashed?

[Two other possible solutions to the front-running problem: 1. Include proper front-running protection like we did in the Legendre bounty contract (probably don't want the complexity); 2. Give a reward only to the block producer]

Reply:
The incentives of the policing of the game are much cleaner when the natural policemen can be directly rewarded. If we only reward the proposer, then the attester-policemen who likely found the issue have to make a decision to either (a) wait until they are a proposer to get the reward or (b) broadcast on gossip immediately and forgo the reward for the health of the network.

In the normal case in which the crosslink is actually available, (b) isn't so strong because the network is currently healthy and getting the slashing on chain only serves to dissuade unhealthy activity in the future.

If attesters get rewarded, they are incentivized to broadcast immediately to a public gossip channel such that they can get rewarded and whoever picks it up gets something too.

-------

Agree that some sort of commit reveal or time-based mechanism with multiple on-chain messages should likely be avoided due to complexity.

-------
> if several attesters produced incorrect custody bits, then some of them could collect the whistleblower rewards before being slashed?

This is a slightly weird side effect, but because of the `WHISTLEBLOWER_REWARD_QUOTIENT`, even if all attesters were slashed and received some marginal reward, they would all strictly be losing `1 - 1 / WHISTLEBLOWER_REWARD_QUOTIENT` fraction of balance which is the majority of their capital.